### DISCUSSION TITLE
config.h

### DISCUSSION THREAD

**Discussion by Vollstrecker** [pupnp/pupnp#381] [Q&A] [answered]
While working on the warnings with the posix names I stumpled across config.h and I don't get the meaning. I'm mean I like the idea of controlling features at runtime, but this seems to do that at compile-time.

I see there no point at all. If the feature is enabled, the source for it is listed for compilation, and if the feature is not enabled, the source is not compiled at all. So even when a user specifies a different value later, the lib will not be altered. Autotools doesn't work well when chainbuilding a lib, and cmake just does the described above no matter if standalone or not.

So is there any point of carrying 500+ lines of code that do nothing?

**Comment by mrjimenez** [comment]
Hi Vollstrecker,

If I recall correctly, this is legacy from a time when a few kbytes of memory made a huge difference. So, the idea was to save as much memory as possible, according to the application needs. For example, most of the time you don't need both the UPnP client and server at the same time, so it made a lot of sense removing the unnecessary code.

**Comment by Vollstrecker** [comment]
I get the concept of disabling feature that are not needed, but I don't get how the size can be influenced by deselecting options when compiling the consuming app. If the consumer disables a feature, the code will still be in the compiled lib. This could only work, if the consumer would import the source in their tree and ship the own version of the lib (or link it statically into their app), but that would cause even more impact if I have 2 progs on my system which do this.

Anyway, if you say legacy, I guess I can savely remove this.

**Comment by mrjimenez** [comment]
I think it would be better to leave it, since this can still be usefull on some low memory controllers. There are STM32 MCUs with as low as 64 k (that is right, k) bytes of flash memory.

**Comment by Vollstrecker** [comment]
Sure this would be helpful, I just don't understand how this can influence the size, if the lib is already compiled. I see only 2 ways of using the lib:

1: If the lib is compiled with ENABLE_WEBSERVER 1, then it won't change what later is defined in that file, as the lib is already done.
2: If the consumer copies the source and compiles it into his app, why should he include webserver.c and turn it off on the config instead of just not listing webserver.c as source?

**Comment by mrjimenez** [comment]
I am not sure I understand you, but let me try.

For example, the user of an embedded device will compile libupnp from the sources. This will determine its size. He wants it to have the lowest memory fingerprint possible. He will not import parts of the library source into his project, as it would be more cumbersome and require a deeper comprehension about the inner workings of the library. After all, the point of having a library is to have an API. Then he will compile his application using this API and link to a smaller size library.

It is as if there are, e.g., three incarnations of the library, with different sizes: one with just the client part, one with just the server part and one with both. This does not make sense in a public linux distribution, since the packages are all pre-compiled and not tailored to a specific use. This only makes sense in an embedded environment.

Please, forgive me if I still did not understood your point.

**Comment by Vollstrecker** [comment]
Forgiven, but we're getting closer. So it's all about creating the lib with different features, and the consumer compiles the lib and later links against it. If this is the the workflow you refer to, the user can use configure --disable-client or cmake -DUPNP_ENABLE_CLIENT_API=FALSE, so the client api code is not compiled into the lib and the lib is smaller for the smaller environment.

But both autotools and cmake do this by defining UPNP_HAVE_CLIENT in autoconf.h this get's included by config.h and if UPNP_HAVE_CLIENT is defined INCLUDE_CLIENT_APIS get's defined to 1. Later in the prog there's #ifdef INCLUDE_CLIENT_APIS, if we just replace this with #ifdef UPNP_HAVE_CLIENT the intermediate step can completely be deleted.

For ssdp this goes further as it defines to 0 or 1 and inverts the value of UPNP_HAVE_SSDP and in the complete source only for == 0 is checked. So changing these #if EXCLUDE_SSDP == 0 to #ifdef UPNP_HAVE_SSDP would just make another 7 lines obsolete.

**Comment by Vollstrecker** [comment]
Not to mention the webserver-part in config.h which adds sanaty checks that are already done in the config systems.

What stays is the definition of THREAD_IDLE_TIME and so on. With that I'm unsure, I didn't check if they are each only needed at specific places where a corresponding header can be found where it would be better defined near it's usage, otoh it's nice to have a central point where all defs go.

**Comment by mrjimenez** [comment, marked as answer]
Oh, I see what you mean. As long as you can remove the unneeded code and have the option to have a smaller library, I see no point in keeping these intermediate steps.

The original code was a lot messier. There were ugly macros spread out around the code and no sane person was able to read it, not to mention reformat the code to a proper standard, since it was far from being C language. I believe that these are the remains of these changes, I am totally ok with removing the redundant and useless ifdefs.
